src/work/Work.cpp

```diff
@@ -4,434 +4,224 @@
     {
-    case WORK_COMPLETE_OK:
-        setState(onSuccess());
-        break;
-    case WORK_COMPLETE_FAILURE:
-        setState(WORK_FAILURE_RETRY);
-        break;
-    case WORK_COMPLETE_FATAL:
-        setState(WORK_FAILURE_FATAL);
-        break;
-    }
-
-    switch (getState())
-    {
-    case WORK_SUCCESS:
-        succ.Mark();
-        CLOG(DEBUG, "Work")
-            << "notifying parent of successful " << getUniqueName();
-        notifyParent();
-        break;
-
-    case WORK_FAILURE_RETRY:
-        fail.Mark();
-        onFailureRetry();
-        scheduleRetry();
-        break;
-
-    case WORK_FAILURE_RAISE:
-    case WORK_FAILURE_FATAL:
-        fail.Mark();
-        onFailureRaise();
-        CLOG(DEBUG, "Work") << "notifying parent of failed " << getUniqueName();
-        notifyParent();
-        break;
-
-    case WORK_PENDING:
-        succ.Mark();
-        advance();
-        break;
-
-    case WORK_RUNNING:
-        succ.Mark();
-        scheduleRun();
-        break;
-
-    default:
-        assert(false);
-        break;
+        mNextChild = mChildren.begin();
     }
 }
 
-void
-Work::onReset()
+bool
+Work::allChildrenSuccessful() const
 {
+    return std::all_of(mChildren.begin(), mChildren.end(),
+                       [](std::shared_ptr<BasicWork> const& w) {
+                           return w->getState() ==
+                                  BasicWork::State::WORK_SUCCESS;
+                       });
 }
 
-void
-Work::onStart()
+bool
+Work::allChildrenDone() const
 {
+    return std::all_of(
+        mChildren.begin(), mChildren.end(),
+        [](std::shared_ptr<BasicWork> const& w) { return w->isDone(); });
 }
 
-void
-Work::onRun()
+bool
+Work::anyChildRunning() const
 {
-    scheduleSuccess();
+    return std::any_of(mChildren.begin(), mChildren.end(),
+                       [](std::shared_ptr<BasicWork> const& w) {
+                           return w->getState() ==
+                                  BasicWork::State::WORK_RUNNING;
+                       });
 }
 
-Work::State
-Work::onSuccess()
+bool
+Work::hasChildren() const
 {
-    return WORK_SUCCESS;
+    return !mChildren.empty();
 }
 
-void
-Work::onFailureRetry()
+bool
+Work::anyChildRaiseFailure() const
 {
+    return std::any_of(mChildren.begin(), mChildren.end(),
+                       [](std::shared_ptr<BasicWork> const& w) {
+                           return w->getState() ==
+                                  BasicWork::State::WORK_FAILURE;
+                       });
 }
 
-void
-Work::onFailureRaise()
+std::shared_ptr<BasicWork>
+Work::yieldNextRunningChild()
 {
-}
+    while (mNextChild != mChildren.end())
+    {
+        auto next = mNextChild;
+        mNextChild++;
+        assert(*next);
+        auto state = (*next)->getState();
+        if (state == State::WORK_RUNNING)
+        {
+            return *next;
+        }
+        else if (state == State::WORK_SUCCESS)
+        {
+            mNextChild = mChildren.erase(next);
```

Comment:
I did not remove works that failed, because some works want to wait for all children to complete before deciding what status to return; However, I think it's also possible to remove failed  children as well and just add more checks to the derived classes. 

I think this is related to a few other comments below. To summarize my thoughts, I think the idea is: on each iteration, parent checks the status of its children; it might decide to complete (maybe one child failed, causing parent to fail as well). It might decide to continue running. On the next iteration, completed children are cleared up during round robin, which is okay because parent already knows the state of removed children (FWIW, in the vast majority of cases today as soon the child fails, parent fails too; otherwise, parent waits for all children to succeed) Additionally, I think we should enforce parent to only return success if all of its children are successful. Let me know what you think @MonsieurNicolas